filesetup.c

```diff
@@ -833,12 +833,26 @@ static unsigned long long get_fs_free_counts(struct thread_data *td)
 uint64_t get_start_offset(struct thread_data *td, struct fio_file *f)
 {
 	struct thread_options *o = &td->o;
+	unsigned long long offset;
 
 	if (o->file_append && f->filetype == FIO_TYPE_FILE)
 		return f->real_file_size;
 
-	return td->o.start_offset +
-		td->subjob_number * td->o.offset_increment;
+	if (o->start_offset_percent > 0) {
+		// calculate the raw offset
+		offset = (f->real_file_size * o->start_offset_percent / 100) +
+			(td->subjob_number * td->o.offset_increment);
+
+		// block align the offset at the next available boundary
+		offset = (offset / o->ba[o->td_ddir] +
+			(offset % o->ba[o->td_ddir] != 0))  // ceiling(offset/blockalignment)
```

Comment:
I'm not sure you can use `o->td_ddir` to index into `o->ba`. Isn't o->ba[] only 3 big (DDIR_RWDIR_CNT 
 comes from https://github.com/axboe/fio/blob/c1f50f765a0a51931605c1fb223d166e3b3a93c6/io_ddir.h#L8 and ba is defined in https://github.com/axboe/fio/blob/master/thread_options.h#L81 ) but o->td_ddir is kind of a bitfield (see https://github.com/axboe/fio/blob/c1f50f765a0a51931605c1fb223d166e3b3a93c6/io_ddir.h#L29 ).